### ISSUE TITLE
with multimaster not detected read only nodes correctly via config mode

### ISSUE THREAD

**Issue by blackluck** [signal18/replication-manager#112]
The 'multimaster = true' option not working if set in configuration file.
It looks like not read that option from the config file, because otherwise the config works if start with 'replication-manager monitor --multimaster' .

**Comment by tanji** [comment]
Hi,

what version do you use?

I just made the following test:
```
# echo "multimaster = true" >> /etc/replication-manager/config.toml
# replication-manager monitor --verbose --daemon
2017/01/24 17:13:17 INFO : Using config file: /etc/replication-manager/config.toml
2017/01/24 17:13:17 {User:root:bb8460645aed6d7e511ddc9bc1dd0aabba41d667a3 Hosts:172.16.1.1,172.16.1.2,172.16.1.3 Socket: RplUser:root:bb8460645aed6d7e511ddc9bc1dd0aabba41d667a3 Interactive:true Verbose:true PreScript: PostScript: MaxDelay:10 GtidCheck:false PrefMaster: IgnoreSrv: WaitKill:5000 WaitTrx:15 WaitWrite:10 ReadOnly:true MaxFail:5 FailResetTime:300 Autorejoin:true LogFile:/tmp/mrmd.log MonitoringTicker:2 Timeout:1 FailLimit:5 FailTime:0 CheckType:tcp CheckReplFilter:true CheckBinFilter:true RplChecks:true MasterConn: MultiMaster:true Spider:false BindAddr:localhost HttpPort:9999 HttpServ:true HttpRoot:/home/tanj/Dev/go/src/github.com/tanji/replication-manager/dashboard HttpAuth:false HttpBootstrapButton:false SessionLifeTime:3600 Daemon:true MailFrom:mrm@localhost MailTo: MailSMTPAddr:localhost:25 MasterConnectRetry:10 FailSync:false FailEventScheduler:false FailEventStatus:false Heartbeat:false MxsOn:false MxsHost:127.0.0.1 MxsPort:6603 MxsUser:admin MxsPass:mariadb HaproxyOn:false HaproxyWritePort:3306 HaproxyReadPort:3307 HaproxyWriteBindIp:0.0.0.0 HaproxyReadBindIp:0.0.0.0 HaproxyBinaryPath:/usr/sbin/haproxy KeyPath:/etc/replication-manager/.replication-manager.key LogLevel:2 Test:false}
```

As you can see in the config debug output above, multimaster is set to true.

**Comment by blackluck** [comment]
```
# replication-manager version
MariaDB Replication Manager version 0.7
Version:  0.7.0-rc3
Build Time:  2016-11-23T14:38:23+0100
```
(downloaded as deb package)

For me it didn't say with verbose the configuration only this:

```
# grep "multimaster" /etc/replication-manager/config.toml
multimaster = true
```
```
# replication-manager monitor --verbose --daemon
2017/01/24 17:51:04 INFO : replication-manager version 0.7 started in daemon mode
2017/01/24 17:51:04 INFO : Monitor started in automatic mode
2017/01/24 17:51:04 INFO : Starting http monitor on port 10001
2017/01/24 17:51:06 WARNING: ERR00011 Multiple masters were detected
2017/01/24 17:51:06 ERROR: ERR00012 Could not autodetect a master
```

and with the same config but start with --multimaster option:

```
# replication-manager monitor --verbose --daemon --multimaster
2017/01/24 17:53:53 INFO : replication-manager version 0.7 started in daemon mode
2017/01/24 17:53:53 INFO : Monitor started in automatic mode
2017/01/24 17:53:53 INFO : Starting http monitor on port 10001
2017/01/24 17:53:55 WARNING: WARN00004 RO server count > 1 in multi-master mode.  switching to prefered master.
2017/01/24 17:53:55 ERROR: ERR00012 Could not autodetect a master
2017/01/24 17:53:57 WARNING: WARN00004 HAS BEEN FIXED, RO server count > 1 in multi-master mode.  switching to prefered master.
2017/01/24 17:53:57 ERROR: ERR00012 HAS BEEN FIXED, Could not autodetect a master
2017/01/24 17:53:57 WARNING: WARN00004 HAS BEEN FIXED, RO server count > 1 in multi-master mode.  switching to prefered master.
2017/01/24 17:53:57 ERROR: ERR00012 HAS BEEN FIXED, Could not autodetect a master
2017/01/24 17:53:57 WARN : WARN00007 At least one server is not ACID-compliant. Please check that the values of sync_binlog and innodb_flush_log_at_trx_commit are set to 1
```
Last message is normal, sync_binlog and innodb_flush_log_at_trx aren't 1 and after that it's working. But if multimaster only in config file, then it's not.

the probably relevant config parts are (ip masked):
hosts = "db1:3307,db2:3307"
connect-timeout = 1
prefmaster = "db1"
multimaster = true
autorejoin = true
readonly = true
interactive = false
wait-kill = 5000
post-failover-script = "/usr/local/sbin/mysql-vipup.sh"
pre-failover-script = "/usr/local/sbin/mysql-vipdown.sh"

**Comment by svaroqui** [comment]
2017/01/24 17:51:06 Multiple masters were detected
2017/01/24 17:51:06 ERROR: ERR00012 Could not autodetect a master
 replication manager run in multi master so issue title may be wrong the question is why in config mode it does not detect both nodes in read only like in command line mode

**Comment by blackluck** [comment]
It wasn't clear for me, modified title.

Tried with log-level = 3 and see this:

```
# replication-manager monitor --verbose --daemon
2017/01/24 18:59:36 INFO : replication-manager version 0.7 started in daemon mode
2017/01/24 18:59:36 INFO : Monitor started in automatic mode
2017/01/24 18:59:36 INFO : Starting http monitor on port 10001
2017/01/24 18:59:38 DEBUG: Discovering topology loop
2017/01/24 18:59:38 DEBUG: Entering topology detection
2017/01/24 18:59:38 DEBUG: Server db1:3307 is configured as a slave
2017/01/24 18:59:38 DEBUG: Privilege check on db1:3307
2017/01/24 18:59:38 DEBUG: Server db2:3307 is configured as a slave
2017/01/24 18:59:38 DEBUG: Privilege check on db2:3307
2017/01/24 18:59:38 WARNING: ERR00011 Multiple masters were detected
2017/01/24 18:59:38 ERROR: ERR00012 Could not autodetect a master
2017/01/24 18:59:40 DEBUG: Discovering topology loop
2017/01/24 18:59:40 DEBUG: Entering topology detection
2017/01/24 18:59:40 DEBUG: Server db1:3307 is configured as a slave
2017/01/24 18:59:40 DEBUG: Privilege check on db1:3307
2017/01/24 18:59:40 DEBUG: Server db2:3307 is configured as a slave
2017/01/24 18:59:40 DEBUG: Privilege check on db2:3307
2017/01/24 18:59:42 DEBUG: Discovering topology loop
2017/01/24 18:59:42 DEBUG: Entering topology detection
2017/01/24 18:59:42 DEBUG: Server db1:3307 is configured as a slave
2017/01/24 18:59:42 DEBUG: Privilege check on db1:3307
2017/01/24 18:59:42 DEBUG: Server db2:3307 is configured as a slave
2017/01/24 18:59:42 DEBUG: Privilege check on db2:3307
```
and it continue the same.

How could I get more detail or what should I check to see why it's not working?

**Comment by tanji** [comment]
OK, that's weird. Your version should have the line that prints "Using config file:" 
If it doesn't, it could mean that the config file does not exist or could not be read:
```
$ sudo chown 600 /etc/replication-manager/config.toml
$ ./replication-manager monitor --log-level=2 --daemon
2017/01/24 20:36:40 {User: Hosts: Socket: RplUser: Interactive:true Verbose:false PreScript: PostScript: MaxDelay:0 GtidCheck:false PrefMaster: IgnoreSrv: WaitKill:5000 WaitTrx:10 WaitWrite:10 ReadOnly:true MaxFail:5 FailResetTime:300 Autorejoin:true LogFile: MonitoringTicker:2 Timeout:5 FailLimit:5 FailTime:0 CheckType:tcp CheckReplFilter:true CheckBinFilter:true RplChecks:true MasterConn: MultiMaster:false Spider:false BindAddr:localhost HttpPort:10001 HttpServ:false HttpRoot:/usr/share/replication-manager/dashboard HttpAuth:false HttpBootstrapButton:false SessionLifeTime:3600 Daemon:true MailFrom:mrm@localhost MailTo: MailSMTPAddr:localhost:25 MasterConnectRetry:10 FailSync:false FailEventScheduler:false FailEventStatus:false Heartbeat:false MxsOn:false MxsHost:127.0.0.1 MxsPort:6603 MxsUser:admin MxsPass:mariadb HaproxyOn:false HaproxyWritePort:3306 HaproxyReadPort:3307 HaproxyWriteBindIp:0.0.0.0 HaproxyReadBindIp:0.0.0.0 HaproxyBinaryPath:/usr/sbin/haproxy KeyPath:/etc/replication-manager/.replication-manager.key LogLevel:2 Test:false}
2017/01/24 20:36:40 ERROR: No hosts list specified.
```
But still, the defaults would be printed. So, it really seems your setup has an issue reading the config file.

Can you retry with today's build (tentative release) and see if logging improves?
https://storage.googleapis.com/replication-manager/replication-manager_0.7.0-05d8ee1_amd64.deb
Thanks!

**Comment by blackluck** [comment]
Still it used that config file, only that file contained host and login informations :).

Tried it with that today build and if set log-level then printed out config at start and it worked with multimaster in config file.
So probably it was fixed between today and that release just I didn't found a comment like fix for these issue.

On the other hand maybe found another issue, if set log-level greater then 2 (nowhere a hint about it's numbers) with this today build then break with segfault after this:
```
2017/01/24 22:59:12 INFO : replication-manager version 0.7 started in daemon mode
2017/01/24 22:59:12 INFO : Monitor started in automatic mode
2017/01/24 22:59:12 INFO : Starting http monitor on port 10001
2017/01/24 22:59:14 DEBUG: Monitoring server loop
2017/01/24 22:59:14 DEBUG: Server [0]: URL: db1:3307 State:  Slave PrevState:
2017/01/24 22:59:14 DEBUG: Server [1]: URL: db2:3307 State:  Slave PrevState:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x427888]

goroutine 1 [running]:
panic(0x91a560, 0xc4200120b0)
        /usr/lib/go/src/runtime/panic.go:500 +0x1a1
main.glob..func12(0xc3c680, 0xc420271260, 0x0, 0x2)
        /home/tanj/Dev/go/src/github.com/tanji/replication-manager/repmgr.go:314 +0xac8
github.com/tanji/replication-manager/vendor/github.com/spf13/cobra.(*Command).execute(0xc3c680, 0xc420271100, 0x2, 0x2, 0xc3c680, 0xc420271100)
        /home/tanj/Dev/go/src/github.com/tanji/replication-manager/vendor/github.com/spf13/cobra/command.go:636 +0x443
github.com/tanji/replication-manager/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc3b580, 0x0, 0xc441c8, 0x9a475b)
        /home/tanj/Dev/go/src/github.com/tanji/replication-manager/vendor/github.com/spf13/cobra/command.go:722 +0x367
github.com/tanji/replication-manager/vendor/github.com/spf13/cobra.(*Command).Execute(0xc3b580, 0x0, 0x0)
        /home/tanj/Dev/go/src/github.com/tanji/replication-manager/vendor/github.com/spf13/cobra/command.go:681 +0x2b
main.main()
        /home/tanj/Dev/go/src/github.com/tanji/replication-manager/main.go:87 +0x31
```

Maybe this needs another ticket.

**Comment by tanji** [comment]
@blackluck thanks for the report, if it's simple enough I will hotfix this.

**Comment by tanji** [comment]
Fixed in fb7da86
